src/Features/Core/Portable/TodoComments/AbstractTodoCommentsIncrementalAnalyzer.cs

```diff
@@ -19,6 +20,12 @@ internal abstract partial class AbstractTodoCommentsIncrementalAnalyzer : Increm
         private string? _lastOptionText = null;
         private ImmutableArray<TodoCommentDescriptor> _lastDescriptors = default;
 
+        /// <summary>
+        /// Set of documents that we have reported an empty set of todo comments for.  Don't both re-reporting these
+        /// documents as long as we keep getting no todo comments produced for them.
+        /// </summary>
+        private readonly HashSet<DocumentId> _documentsWithNoTodoComments = new();
```

Comment:
❔ Will this grow forever in SG scenarios where files are generated each time with new IDs?

Reply:
@jasonmalinowski how do SG files work?  do they get fresh DocIds?  or do you reuse them?

also, i can at least remov ethese once we get the notification that the doc is gone.

Reply:
💭 Can we keep track instead of documents that _do_ have comments? This would be a smaller list.

Reply:
we now remove from teh set as well as add to it.

Reply:
> Can we keep track instead of documents that do have comments?

yup.  flipped.

Reply:
> @jasonmalinowski how do SG files work? do they get fresh DocIds? or do you reuse them?

Just to state it directly: source generated documents reuse IDs as long as the generator is giving stable hint names. (That's why we added that API.)

Reply:
good to know.  thanks!